[usm] fix the script generating the gotls/lookup/luts.go file

andreimatei opened the pull request:
Before this patch, the generation of the `gotls/lookup/luts.go` file was
failing: Go versions below 1.16 cannot parse the go.mod file of the
x/net module at versions above 0.35 (released a few months ago). This
patch pins the module at v0.35. We were arbitrarily using the latest
x/net version (and thus using the debug info for the latest); now we're
arbitrarily using a fixed version. Luckily, the debug info we need for
x/net is stable -- we need the offset of an embedded field in a struct.

The regenerated `luts.go` doesn't have any changes. FWIW, I think this shows
that the debug info in that file is valid for Go 1.24; I'm not sure we
ever ran the script for 1.24, since I think the release of x/net that
broke it happened a little before 1.24.

andreimatei committed:
[usm] fix comment

A comment was referencing the wrong type name.

andreimatei committed:
[usm] fix generation script for debug info lookup table

Before this patch, the generation of the gotls/lookup/luts.go file was
failing: Go versions below 1.16 cannot parse the go.mod file of the
x/net module at versions above 0.35 (released a few months ago). This
patch pins the module at v0.35. We were arbitrarily using the latest
x/net version (and thus using the debug info for the latest); now we're
arbitrarily using a fixed version. Luckily, the debug info we need for
x/net is stable -- we need the offset of an embedded field in a struct
(*).

In addition to failing on Go 1.16, the latest version of x/net was
forcing the toolchain selection mechanism in go 1.21+ to select 1.23 --
so we were not actually using go 1.21 and 1.22. The next commit makes
the script robust to this.

The regenerated luts.go doesn't have any changes. FWIW, I think this shows
that the debug info in that file is valid for Go 1.24; I'm not sure we
ever ran the script for 1.24, since I think the release of x/net that
broke it happened a little before 1.24.

andreimatei committed:
[usm] inhibit Go toolchain selection when generating luts.go

The generator script for luts.go intends to use every Go compiler
version. Until this patch, it was fragile because it allowed Go's
toolchain selection mechanism to transparently use a newer toolchain
version than the one the script was intended to use, subject to the
requirements of the modules used by the test program. This could lead to
the generated source containing debug info for the wrong compiler
versions. In fact, this was happening until the previous commit.

This patch makes the generation robust by forcing the intended toolchain
versions.

guyarb commented:
> I'm not sure we
ever ran the script for 1.24, since I think the release of x/net that
broke it happened a little before 1.24.

We did run it. Go 1.24 was released on February 11th, while `x/net` v0.36 (the version that broke the support) was released only on March 4th

guyarb reviewed (changes requested):
reviewed

andreimatei commented on pkg/network/go/lutgen/run.go:
```diff
@@ -324,12 +328,25 @@ func setupGoModule(ctx context.Context, cmd *exec.Cmd, programPath string) error
 	// modify original `exec.Cmd` object by setting the `Dir` field to the one we created
 	cmd.Dir = moduleDir
 
+	// Pin the golang.org/x/net module to an old version. Newer versions cannot
+	// be processed by Go <= 1.16 because the go.mod in x/net has the wrong
+	// format. Newer versions of the package have a go.mod file that can't be
+	// parsed by Go <= 1.16.
+	getCmd := exec.CommandContext(ctx, "go", "get", "golang.org/x/net@v0.35.0")
+	getCmd.Env = cmd.Env
+	getCmd.Dir = cmd.Dir
+	getCmd.Path = cmd.Path
+	output, err := getCmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("error executing 'go get': %s\n%s", err, output)
+	}
```
I've considered doing that, but this all leads to a deeper question -- which version of the library do we care about? Why would would we particularly care about the latest (and only the latest)? Then, caring about different library versions for different compiler versions seems confusing -- although you can argue that it makes some sense since new versions of the lib don't work with old compilers... I'm in the live debugger team, and so the real answer is perhaps that we should actually look at the particular binary and dynamically analyze its debug info :).

But also you can argue that dealing with the debug info for the `x/net` library at all is kinda silly -- the only thing we look at is the offset of an embedded field into a struct -- which seems likely to stay 0 for as long as that field exists. That's why I thought that the complexity of dealing with multiple versions is not really worth it. I guess using the latest version of the library tells us that the embedded field continues to exist, so there's some value in it...

Having written all this, if you think the complexity of changing the library version based on the compiler version is worth it, I'm happy to do it.

andreimatei replied:
I've made the script use different versions of the library based on the Go version, as you suggested.

andreimatei committed:
[usm] add max go version flag to lookup table generation script

For development you might want to run a single Go version. The script
already lets you specify a minimum version.
